server/src/main/java/com/thoughtworks/go/server/service/MaterialService.java

```diff
@@ -169,4 +172,44 @@ private void resolveSecretParams(Material material) {
                 .stream()
                 .collect(toMap(mod -> mod.getMaterialInstance().getFingerprint(), mod -> mod));
     }
+
+    public List<Modification> getModificationsFor(MaterialConfig materialConfig, long afterCursor, long beforeCursor, Integer pageSize) {
+        MaterialInstance materialInstance = materialRepository.findMaterialInstance(materialConfig);
```

Comment:
This query is cached. Also, IMO, `join` would be an expensive operation. Will deploy this as is on build.gocd. Based on the response there - we can later on raise another PR for performance improvement. WDYT?

Reply:
Previously I had seen 3 queries fired to fetch the modifications, hence I suggested we use joins. Since MaterialInstance lookup is cached a query to fetch modifications should be fine. Either way we should benchmark  these queries before taking a decision.